src/cobra/util/array.py

```diff
@@ -69,40 +78,45 @@ def create_stoichiometric_matrix(model, array_type="dense", dtype=None):
         return array
 
 
-def nullspace(A, atol=1e-13, rtol=0):
-    """Compute an approximate basis for the nullspace of A.
-    The algorithm used by this function is based on the singular value
-    decomposition of `A`.
+def nullspace(A: np.ndarray, atol: float = 1E-13, rtol: float = 0.0) -> np.ndarray:
+    r"""Compute an approximate basis for the nullspace of A.
+
+    The algorithm used by this function is based on the Singular Value
+    Decomposition (SVD) of `A`.
 
     Parameters
     ----------
     A : numpy.ndarray
-        A should be at most 2-D.  A 1-D array with length k will be treated
-        as a 2-D with shape (1, k)
-    atol : float
-        The absolute tolerance for a zero singular value.  Singular values
-        smaller than `atol` are considered to be zero.
-    rtol : float
-        The relative tolerance.  Singular values less than rtol*smax are
-        considered to be zero, where smax is the largest singular value.
-
-    If both `atol` and `rtol` are positive, the combined tolerance is the
-    maximum of the two; that is::
-    tol = max(atol, rtol * smax)
-    Singular values smaller than `tol` are considered to be zero.
+        `A` should be at most 2-D. 1-D array with length k will be treated
+        as a 2-D with shape (1, k).
+    atol : float, optional
+        The absolute tolerance for a zero singular value. Singular values
+        smaller than `atol` are considered to be zero (default 1E-13).
+    rtol : float, optional
+        The relative tolerance. Singular values less than `rtol * smax` are
+        considered to be zero, where `smax` is the largest singular value
+        (default 0.0).
 
     Returns
     -------
     numpy.ndarray
         If `A` is an array with shape (m, k), then `ns` will be an array
-        with shape (k, n), where n is the estimated dimension of the
+        with shape (k, n), where `n` is the estimated dimension of the
         nullspace of `A`.  The columns of `ns` are a basis for the
         nullspace; each element in numpy.dot(A, ns) will be approximately
         zero.
 
     Notes
     -----
-    Taken from the numpy cookbook.
+    This is taken from the numpy cookbook.
+
+    If both `atol` and `rtol` are positive, the combined tolerance is the
+    maximum of the two; that is:
+
+    .. math:: \mathtt{tol} = \max(\mathtt{atol}, \mathtt{rtol} * \mathtt{smax})
+
+    Singular values smaller than `tol` are considered to be zero.
+
     """
     A = np.atleast_2d(A)
     u, s, vh = np.linalg.svd(A)
```

Comment:
```suggestion
    _, s, vh = np.linalg.svd(A)
```

Reply:
Scipy has a `nullspace` [implementation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.null_space.html) and I was wondering if we should maintain cobrapy's implementation. I don't see any cobrapy-specific thing being done.

Reply:
If you look at the source code, it's very similar. https://github.com/scipy/scipy/blob/v1.5.2/scipy/linalg/decomp_svd.py#L330-L388 Let's not introduce a scipy dependence here but if you like you can adjust the code of the function to match the scipy implementation.

Reply:
The functionality doesn't change to a large extent so I guess it's okay to leave it as it is.